File: 1000-9999/1525.cpp
// 1525번 퍼즐

#include <bits/stdc++.h>
using namespace std;

int board, ans=-1;
unordered_map<int, bool> check;

int change[4]={1, -1, 3, -3};

int lpow(int a, int n){
    int res=1;
    for(int i=0; i<n; i++) res*=a;
    return res;
}

int move(int board, int l, int l2){
    l=8-l; l2=8-l2;
    int tmp=board/lpow(10, l)%10;
    int tmp2=board/lpow(10, l2)%10;

    board-=lpow(10, l)*tmp;
    board+=lpow(10, l)*tmp2;
    board-=lpow(10, l2)*tmp2;
    board+=lpow(10, l2)*tmp;

    return board;
}

void solve(int cnt, int board, int l){
    if(cnt==2){
        cout<<board<<"\n";
    }
    if(ans!=-1) return;
    if(board==123405786){
        ans=cnt;
        return;
    }
    for(int i=0; i<4; i++){
        int L=l+change[i];
        if(L>=0 && L<9){
            if((l%3==2 && i==0) || (l%3==0 && i==1)) continue; 
            int b=move(board, l, L);
            if(!check[b]){
                check[b]=true;
                solve(cnt+1, b, L);
            }
        }
    }
}

int main(void){
    ios::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    
    int b, c;
    for(int i=0; i<9; i++){
        cin>>b;
        if(b==0){
            c=i; b=9;
        }
        board*=10; board+=b;
    }

    solve(0, board, c);
    cout<<ans<<"\n";
}

File: 1000-9999/1525.py
'''
1525번
퍼즐
'''

import sys
from collections import deque
input=sys.stdin.readline

board=0; z=0
for i in range(3):
    a=list(map(int, input().split()))
    for j in range(3):
        if a[j]==0: 
            a[j]=9
            z=(i*3)+j
        board=board*10+a[j]

change=[1, -1, 3, -3]
visited={}; ans=-1

def move(b, l, l2):
    l=8-l; l2=8-l2
    tmp=b//pow(10, l)%10
    tmp2=b//pow(10, l2)%10

    b-=pow(10, l)*tmp
    b+=pow(10, l)*tmp2
    b-=pow(10, l2)*tmp2
    b+=pow(10, l2)*tmp

    return b

q=deque()
q.append((0, board, z))
while q and ans==-1:
    node=q.popleft()
    cnt, board2, l=node
    if board2==123456789: ans=cnt
    for i in range(4):
        L=l+change[i]
        if((L<0 or L>8) or (l%3==2 and i==0) or (l%3==0 and i==1)): continue
        b=move(board2, l, L)
        if b not in visited:
            q.append((cnt+1, b, L))
            visited[b]=True

print(ans)